### DISCUSSION TITLE
Make zsh shell-integration title changing toggleable by scripts after load time

### DISCUSSION THREAD

**Discussion by EightBitBoot** [ghostty-org/ghostty#3659] [Feature Requests, Ideas] [answered]
When changing the current surface's title via ANSI escape escape sequences (`\e]0;My Custom Title Here\a` or `\e]2;My Custom Title Here\a`), the zsh shell integration's `_ghostty_precmd` and `_ghostty_preexec` functions automatically overwrite it (I spent the better part of today tearing my head out and searching through the code thinking this was a bug in the terminal itself). This functionality can be disabled, however it must be disabled _before_ the shell finishes loading and cannot be re-enabled (or disabled) at a later point.

To work around this, a new environment variable has been added (only to the zsh integration for now): `GHOSTTY_SHELL_INTEGRATION_TOGGLE_TITLE_OFF`. This is settable by scripts and commands to disable the zsh integration's overwriting of the title and can be unset to re-enable the functionality.

To take advantage of this, I have the following functions enabled in my .zshrc:

```zsh
function unset_title() {
    echo -ne "\x1b]0;\x07"
    if [[ $TERM == "xterm-ghostty" ]]; then
        unset GHOSTTY_SHELL_INTEGRATION_TOGGLE_TITLE_OFF
    fi
}

function set_title() {
    if [ $# -eq 0 -o ${#1} -eq 0 ]; then
        # If the title isn't provided or is an empty string, reset the title to factory settings
        unset_title
    else
        echo -ne "\x1b]0;${1}\x07"
        if [[ $TERM == "xterm-ghostty" ]]; then
            GHOSTTY_SHELL_INTEGRATION_TOGGLE_TITLE_OFF=1
        fi
    fi
}
```

After taking a cursory look at the bash I don't believe this same method will be possible there (as it works by setting $PS0 and $PS1 to send the escape sequences).

For elvish I don't know enough to add the desired functionality.

My code is available here: https://github.com/ghostty-org/ghostty/compare/main...EightBitBoot:ghostty:toggle_zsh_title_integration.

**Comment by niklaas** [comment]
I came across this post while desperately finding out why the title of my terminal keeps updating in a way I don't want to. Apparently it's Ghostty's shell integration that does that. For those who are wondering how to disable it, put the following in the configuration file:

```
shell-integration = none
```

**Reply by taylrfnt** [reply]
This is one method, albeit it's sort of the nuclear option.  You can use the following to disable the title setting without losing all the other features of shell integration, if you want:
```ini
shell-integration-features = no-title
```

Ref: https://ghostty.org/docs/config/reference#shell-integration-features

**Reply by niklaas** [reply]
Nice, thanks for the hint!

**Comment by jparise** [comment, marked as answer]
I don't think this is something we'll support as a general, cross-shell feature. I would recommend disabling the `title` shell feature (`shell-integration-features = no-title`) if you're looking for different behavior, which you can implement in your local environment.
